### DISCUSSION TITLE
Log per channel

### DISCUSSION THREAD

**Discussion by dfauraz** [nextgenhealthcare/connect#5393] [Feature/Enhancement Ideas]
It would be extremely useful to have one log per channel, instead of all logs together.
In the dashboard, when you have 300 channels running (or even a dozen), the log is flooded by log messages from everywhere (even if you filter by log level, you depend on others not logging at a too high level, which cannot be guaranteed).
It would be helpful to restrict the logs to the channel you are interested in: connector, script and processing logs all together.
This could be done with different log files per channel, or with one file but an additional "column" with the channel name.
Error that cannot be linked to a channel could be linked to other processes (pruner ...), or just to "server".
Finally, if you could set the log level per channel dynamically, then it would be perfect.
This would be a very valid alternative to a debugger, and likely much easier to implement.
This would tremendously help during troubleshooting and developing.

**Comment by pacmano1** [comment]
That would be helpful. 

Just a couple of notes:  There is a debugger although you have to be local to the server.   And using logger.* calls to debug do litter the console, it's perhaps better to use channel map  or other map variables for debugging and leave the console log for catastrophic errors and error mirth generates automatically.  

I think a "view all" errors would still be needed since no one wants to click through 300 channels to find an error.

**Comment by jonbartels** [comment]
Like this? https://github.com/odoodo/Mirth-MetaAppender

**Reply by pacmano1** [reply]
Totally forgot about that one, thanks @jonbartels!

**Reply by dfauraz** [reply]
Thanks a lot @jonbartels, I will look at it

**Reply by dfauraz** [reply]
@jonbartels do you know if the meta appender is used in production somewhere ?

**Reply by jonbartels** [reply]
@dfauraz I don't know for certain. A good gauge would be to check the project GH issues and search https://forums.mirthproject.io/ . The number of questions may show how many users are using it. 

odoodo has done excellent work. I strongly suspect it's used in a lot of places and "just works" without a lot of extra effort or problems.

**Reply by odoodo** [reply]
We have been using it for years without any issues on 4 production servers w/ about 100-150 channels per server (It comprises also a cumulative error log for convenience). The correspond thread can be found [here ](https://forums.mirthproject.io/forum/mirth-connect/support/18801-new-feature-mirth-logging-replacement)on the Mirth forum.

As we are still on Mirth v3.7.1 I am not sure how well it gets along with the Log4J-upgrade to v2.x in the latest Mirth version. Maybe someone has tested it and could provide some feedback?
